go/client/logclient.go

```diff
@@ -178,44 +179,66 @@ func (c *LogClient) postAndParse(uri string, req interface{}, res interface{}) (
 	return resp, string(body), nil
 }
 
+func backoffForRetry(ctx context.Context, d int) error {
+	backoffTimer := time.NewTimer(time.Duration(d) * time.Second)
+	if ctx != nil {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-backoffTimer.C:
+		}
+	} else {
+		<-backoffTimer.C
+	}
+	return nil
+}
+
 // Attempts to add |chain| to the log, using the api end-point specified by
-// |path|.
-func (c *LogClient) addChainWithRetry(path string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error) {
+// |path|. If provided context expires before submission is complete an
+// error will be returned.
+func (c *LogClient) addChainWithRetry(ctx context.Context, path string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error) {
 	var resp addChainResponse
 	var req addChainRequest
 	for _, link := range chain {
 		req.Chain = append(req.Chain, base64.StdEncoding.EncodeToString(link))
 	}
-	done := false
 	httpStatus := "Unknown"
+	backoffSeconds := 0
+	done := false
 	for !done {
-		backoffSeconds := 0
+		if backoffSeconds > 0 {
+			log.Printf("Got %s, backing-off %d seconds", httpStatus, backoffSeconds)
+		}
+		err := backoffForRetry(ctx, backoffSeconds)
+		if err != nil {
+			return nil, err
+		}
+		if backoffSeconds > 0 {
+			backoffSeconds = 0
+		}
 		httpResp, errorBody, err := c.postAndParse(c.uri+path, &req, &resp)
 		if err != nil {
-			log.Printf("Got %s, backing off.", err)
 			backoffSeconds = 10
-		} else {
-			switch {
-			case httpResp.StatusCode == 200:
-				done = true
-				break
-			case httpResp.StatusCode == 408:
-			case httpResp.StatusCode == 503:
-				// Retry
-				backoffSeconds = 10
-				if retryAfter := httpResp.Header.Get("Retry-After"); retryAfter != "" {
-					if seconds, err := strconv.Atoi(retryAfter); err != nil {
-						backoffSeconds = seconds
-					}
+			continue
+		}
+		switch {
+		case httpResp.StatusCode == 200:
+			done = true
+			break
+		case httpResp.StatusCode == 408:
+			// request timeout, retry immediately
+		case httpResp.StatusCode == 503:
+			// Retry
+			backoffSeconds = 10
+			if retryAfter := httpResp.Header.Get("Retry-After"); retryAfter != "" {
+				if seconds, err := strconv.Atoi(retryAfter); err == nil {
+					backoffSeconds = seconds
 				}
-			default:
-				return nil, fmt.Errorf("Got HTTP Status %s: %s", httpResp.Status, errorBody)
 			}
-			httpStatus = httpResp.Status
+		default:
+			return nil, fmt.Errorf("Got HTTP Status %s: %s", httpResp.Status, errorBody)
```

Comment:
Lowercase "g" and no full-stop on the end of the error msg (see https://github.com/golang/go/wiki/CodeReviewComments#error-strings.)